ati-drivers does not build with kernel 4.6

## Issue description

Two compilation errors are seen.  I have a fix at https://github.com/jerith666/nixpkgs/commit/a629947313995a1554d20a3f203fbd10b91f9afc, but I expect that it needs to be variablized to only apply these two new patches if building against a 4.6 (or greater) kernel, and I don't know how to do that.

add'l keywords: ati_unfree

Compiler output:

```
/tmp/nix-build-ati-drivers-15.12-4.6.7.drv-0/common/lib/modules/fglrx/build_mod/2.6.x/firegl_public.c:3245:11: error: too many arguments to function 'get_user_pages'
     ret = get_user_pages(current, current->mm, vaddr, page_cnt, 0, 0, (struct page **)page_list, NULL);
           ^
In file included from /nix/store/jkp1xn8ma57pi0jzfblf4fmfi9nbbjlp-linux-4.6.7-dev/lib/modules/4.6.7/source/include/linux/scatterlist.h:7:0,
                 from /nix/store/jkp1xn8ma57pi0jzfblf4fmfi9nbbjlp-linux-4.6.7-dev/lib/modules/4.6.7/source/include/linux/dmapool.h:14,
                 from /nix/store/jkp1xn8ma57pi0jzfblf4fmfi9nbbjlp-linux-4.6.7-dev/lib/modules/4.6.7/source/include/linux/pci.h:1242,
                 from /tmp/nix-build-ati-drivers-15.12-4.6.7.drv-0/common/lib/modules/fglrx/build_mod/2.6.x/firegl_public.c:119:
/nix/store/jkp1xn8ma57pi0jzfblf4fmfi9nbbjlp-linux-4.6.7-dev/lib/modules/4.6.7/source/include/linux/mm.h:1266:6: note: declared here
 long get_user_pages(unsigned long start, unsigned long nr_pages,
      ^
/tmp/nix-build-ati-drivers-15.12-4.6.7.drv-0/common/lib/modules/fglrx/build_mod/2.6.x/firegl_public.c: In function 'KCL_UnlockUserPages':
/tmp/nix-build-ati-drivers-15.12-4.6.7.drv-0/common/lib/modules/fglrx/build_mod/2.6.x/firegl_public.c:3256:9: error: implicit declaration of function 'page_cache_release' [-Werror=implicit-function-declaration]
         page_cache_release((struct page*)page_list[i]);
         ^
```

--- a/pkgs/os-specific/linux/ati-drivers/default.nix
+++ b/pkgs/os-specific/linux/ati-drivers/default.nix
@@ -71,8 +71,11 @@ stdenv.mkDerivation rec {
     ./patches/15.9-kcl_str.patch
     ./patches/15.9-mtrr.patch
     ./patches/15.9-preempt.patch
-    ./patches/15.9-sep_printf.patch
-  ];
+    ./patches/15.9-sep_printf.patch ]
+  ++ optionals ( kernel != null &&
+                 (builtins.compareVersions kernel.version "4.6") >= 0 )
+               [ ./patches/kernel-4.6-get_user_pages.patch
+                 ./patches/kernel-4.6-page_cache_release-put_page.patch ];
 
   buildInputs =
     [ xorg.libXrender xorg.libXext xorg.libX11 xorg.libXinerama xorg.libSM

--- a/pkgs/os-specific/linux/ati-drivers/patches/kernel-4.6-get_user_pages.patch
+++ b/pkgs/os-specific/linux/ati-drivers/patches/kernel-4.6-get_user_pages.patch
@@ -1,0 +1,25 @@
+diff --git a/common/lib/modules/fglrx/build_mod/firegl_public.c b/common/lib/modules/fglrx/build_mod/firegl_public.c
+index 9c70211..b2242af 100755
+--- a/common/lib/modules/fglrx/build_mod/firegl_public.c
++++ b/common/lib/modules/fglrx/build_mod/firegl_public.c
+@@ -3220,7 +3220,7 @@ int ATI_API_CALL KCL_LockUserPages(unsigned long vaddr, unsigned long* page_list
+     int ret;
+ 
+     down_read(&current->mm->mmap_sem);
+-    ret = get_user_pages(current, current->mm, vaddr, page_cnt, 1, 0, (struct page **)page_list, NULL);
++    ret = get_user_pages(vaddr, page_cnt, 1, 0, (struct page **)page_list, NULL);
+     up_read(&current->mm->mmap_sem);
+ 
+     return ret;
+@@ -3238,7 +3238,7 @@ int ATI_API_CALL KCL_LockReadOnlyUserPages(unsigned long vaddr, unsigned long* p
+     int ret;
+ 
+     down_read(&current->mm->mmap_sem);
+-    ret = get_user_pages(current, current->mm, vaddr, page_cnt, 0, 0, (struct page **)page_list, NULL);
++    ret = get_user_pages(vaddr, page_cnt, 0, 0, (struct page **)page_list, NULL);
+     up_read(&current->mm->mmap_sem);
+ 
+     return ret;
+-- 
+2.9.2
+

--- a/pkgs/os-specific/linux/ati-drivers/patches/kernel-4.6-page_cache_release-put_page.patch
+++ b/pkgs/os-specific/linux/ati-drivers/patches/kernel-4.6-page_cache_release-put_page.patch
@@ -1,0 +1,16 @@
+diff --git a/common/lib/modules/fglrx/build_mod/firegl_public.c b/common/lib/modules/fglrx/build_mod/firegl_public.c
+index b2242af..586129c 100755
+--- a/common/lib/modules/fglrx/build_mod/firegl_public.c
++++ b/common/lib/modules/fglrx/build_mod/firegl_public.c
+@@ -3249,7 +3249,7 @@ void ATI_API_CALL KCL_UnlockUserPages(unsigned long* page_list, unsigned int pag
+     unsigned int i;
+     for (i=0; i<page_cnt; i++)
+     {
+-        page_cache_release((struct page*)page_list[i]);
++        put_page((struct page*)page_list[i]);
+     }
+ }
+ 
+-- 
+2.9.2
+

